TheForceEngine/TFE_DarkForces/player.cpp

```diff
@@ -1830,69 +1830,76 @@ namespace TFE_DarkForces
+				speedDelta = maxFrameChange;
 			}
-			if ((s_playerSector->flags1 & SEC_FLAGS1_ICE_FLOOR) && !s_wearingCleats)
+			else if (speedDelta < -maxFrameChange)
 			{
-				playerSpeed = 0;
+				speedDelta = -maxFrameChange;
 			}
+			s_playerSpeed += speedDelta;
+		}
+		sinCosFixed((s_curTick & 0xffff) << 7, &s_wpnSin, &s_wpnCos);
 
-			if (playerSpeed != s_playerSpeed)
-			{
-				fixed16_16 speedDelta = playerSpeed - s_playerSpeed;
-				fixed16_16 maxFrameChange = mul16(FIXED(32), s_deltaTime);
-
-				if (speedDelta > maxFrameChange)
-				{
-					speedDelta = maxFrameChange;
-				}
-				else if (speedDelta < -maxFrameChange)
-				{
-					speedDelta = -maxFrameChange;
-				}
-				s_playerSpeed += speedDelta;
-			}
-			sinCosFixed((s_curTick & 0xffff) << 7, &s_wpnSin, &s_wpnCos);
-
-			fixed16_16 playerSpeedFract = div16(s_playerSpeed, FIXED(32));
-			s_headwaveVerticalOffset = mul16(mul16(s_wpnCos, PLAYER_HEADWAVE_VERT_SPD), playerSpeedFract);
+		fixed16_16 playerSpeedFract = div16(s_playerSpeed, FIXED(32));
+		headwaveVerticalOffset = mul16(mul16(s_wpnCos, PLAYER_HEADWAVE_VERT_SPD), playerSpeedFract);
 
-			sinCosFixed((s_curTick & 0xffff) << 6, &s_wpnSin, &s_wpnCos);
-			xWpnWaveOffset = mul16(playerSpeedFract, s_wpnCos) >> 12;
+		sinCosFixed((s_curTick & 0xffff) << 6, &s_wpnSin, &s_wpnCos);
+		xWpnWaveOffset = mul16(playerSpeedFract, s_wpnCos) >> 12;
 
-			// Water...
-			if (s_playerSector->secHeight - 1 >= 0)
+		// Water...
+		if (s_playerSector->secHeight - 1 >= 0)
+		{
+			if (s_externalVelX || s_externalVelZ)
 			{
-				if (s_externalVelX || s_externalVelZ)
-				{
-					fixed16_16 externSpd = distApprox(0, 0, s_externalVelX, s_externalVelZ);
-
-					// Replace the fractional part with the current time fractional part.
-					// I think this is meant to add some "randomness" to the headwave while in water.
-					fixed16_16 speed = externSpd & (~0xffff);
-					speed |= (s_curTick & 0xffff);
-					// Then multiply by 16 and take the cosine - this is meant to be a small modification to the weapon motion
-					// (note that the base multiplier is 128).
-					sinCosFixed(speed << 4, &s_wpnSin, &s_wpnCos);
-					// Modify the headwave motion.
-					s_headwaveVerticalOffset += mul16(s_wpnCos, PLAYER_HEADWAVE_VERT_WATER_SPD);
-				}
+				fixed16_16 externSpd = distApprox(0, 0, s_externalVelX, s_externalVelZ);
+
+				// Replace the fractional part with the current time fractional part.
+				// I think this is meant to add some "randomness" to the headwave while in water.
+				fixed16_16 speed = externSpd & (~0xffff);
+				speed |= (s_curTick & 0xffff);
+				// Then multiply by 16 and take the cosine - this is meant to be a small modification to the weapon motion
+				// (note that the base multiplier is 128).
+				sinCosFixed(speed << 4, &s_wpnSin, &s_wpnCos);
+				// Modify the headwave motion.
+				headwaveVerticalOffset += mul16(s_wpnCos, PLAYER_HEADWAVE_VERT_WATER_SPD);
 			}
 		}
-		setCameraOffset(0, s_headwaveVerticalOffset, 0);
 
```

Comment:
The code below is where the condition was moved to and what ultimately separates the head waving and weapon waving, as well as applying the new setting.